### DISCUSSION TITLE
What's the Difference between Unmarshal and Decoder again?

### DISCUSSION THREAD

**Discussion by hollowaykeanho** [pelletier/go-toml#605] [Questions] [answered]
As I gone through https://pkg.go.dev/github.com/pelletier/go-toml/v2#section-readme and building my prototype software, I still can't understand:

1. What's the main difference between `Unmarshal` and `Decoder.Decode` again?
1.1. Both of them did convert TOML data to match Go data structure.
1.2. The difference is not noticeable.
1.3. Same question for their complimentary encoding counterparts.

Reason for asking:

1. Currently building a wrapper package to your `v2` and trying to decide which one to go for. The wrapper would be like:
```
func DecodeFile(path string, structure interface{}) (err error) { ... }
func DecodeData(data string, structure interface{}) (err error) { ... }
```

Since both can achieve the same result, some aspects I might be interested in are:
1. Performance differences.
2. Customization capability.
3. Possible `reflect` package avoidance (I think this is impossible and I assume both uses it to achieve the marshal/unmarshal effect)
4. The path where I can have the least type conversion.
5. Possible of low-memory big data deployment (something like stream cipher in cryptography)

Thanks in adv!

**Comment by moorereason** [comment, marked as answer]
The difference is that `Decoder` and `Encoder` have configuration options.  See the `Set*` methods on those types.

**Reply by hollowaykeanho** [reply]
Noted. In that case, I will use `Decoder` and `Encoder`. Thanks.
